tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,371 @@
+        {
+            var hasArgumentsOrLocals = asyncState.HasLocalsOrReturnValue ||
+                                       asyncState.HasArguments;
+
+            asyncState.HasLocalsOrReturnValue = false;
+            asyncState.HasArguments = false;
+            asyncState.SnapshotCreator.EndEntry(hasArgumentsOrLocals);
+            return asyncState;
+        }
+
+        /// <summary>
+        /// Logs the given <paramref name="arg"/> ByRef.
+        /// </summary>
+        /// <typeparam name="TArg">Type of argument.</typeparam>
+        /// <param name="arg">The argument to be logged.</param>
+        /// <param name="type">The type of the argument</param>
+        /// <param name="paramName">The argument name</param>
+        /// <param name="index">Index of given argument.</param>
+        /// <param name="asyncState">Debugger asyncState</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void LogArg<TArg>(ref TArg arg, Type type, string paramName, int index, ref AsyncMethodDebuggerState asyncState)
+        {
+            asyncState.SnapshotCreator.CaptureArgument(arg, paramName, index == 0, asyncState.HasLocalsOrReturnValue, type);
+            asyncState.HasLocalsOrReturnValue = false;
+            asyncState.HasArguments = true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void LogLocal<TLocal>(ref TLocal local, Type type, string localName, int index, ref AsyncMethodDebuggerState asyncState)
+        {
+            asyncState.SnapshotCreator.CaptureLocal(local, localName, index == 0 && !asyncState.HasLocalsOrReturnValue, type);
+            asyncState.HasLocalsOrReturnValue = true;
+            asyncState.HasArguments = false;
+        }
+
+        /// <summary>
+        /// Logs the given <paramref name="local"/> ByRef.
+        /// </summary>
+        /// <typeparam name="TLocal">Type of local.</typeparam>
+        /// <param name="local">The local to be logged.</param>
+        /// <param name="index">index of given argument.</param>
+        /// <param name="asyncState">Debugger asyncState</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void LogLocal<TLocal>(ref TLocal local, int index, ref AsyncMethodDebuggerState asyncState)
+        {
+            if (!asyncState.IsActive)
+            {
+                return;
+            }
+
+            var localNamesFromPdb = asyncState.MethodMetadataInfo.LocalVariableNames;
+            if (!MethodDebuggerInvoker.TryGetLocalName(index, localNamesFromPdb, out var localName))
+            {
+                return;
+            }
+
+            LogLocal(ref local, typeof(TLocal), localName, index, ref asyncState);
+        }
+
+        /// <summary>
+        /// End Method with Return value invoker - the MoveNext method always return void but here we send the kick-of method return value
```

Comment:
> with Return value

This comment seems to be a copy/paste of the one in the other `EndMethod_StartMarker`, and seems incorrect - isn't this one the one that's meant to handle `async void` methods?

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26